mobile_verifier/src/rewarder.rs

```diff
@@ -0,0 +1,249 @@
+        data_transfer_ingest: FileStore,
+    ) -> Self {
+        Self {
+            pool,
+            reward_period_duration,
+            reward_offset,
+            mobile_rewards,
+            reward_manifests,
+            price_tracker,
+            data_transfer_ingest,
+        }
+    }
+
+    pub async fn run(self, shutdown: triggered::Listener) -> anyhow::Result<()> {
+        loop {
+            let last_rewarded_end_time = self.last_rewarded_end_time().await?;
+            let next_rewarded_end_time = self.next_rewarded_end_time().await?;
+            let scheduler = Scheduler::new(
+                self.reward_period_duration,
+                last_rewarded_end_time,
+                next_rewarded_end_time,
+                self.reward_offset,
+            );
+            let now = Utc::now();
+            let sleep_duration = if scheduler.should_reward(now) {
+                if self.is_data_current(&scheduler.reward_period).await? {
+                    self.reward(&scheduler).await?;
+                    continue;
+                } else {
+                    tracing::info!(
+                        "rewards will be retried in {REWARDS_NOT_CURRENT_DELAY_PERIOD} minutes:"
+                    );
+                    Duration::minutes(REWARDS_NOT_CURRENT_DELAY_PERIOD).to_std()?
+                }
+            } else {
+                scheduler.sleep_duration(now)?
+            };
+
+            tracing::info!(
+                "Sleeping for {}",
+                humantime::format_duration(sleep_duration)
+            );
+
+            tokio::select! {
+                _ = shutdown.clone() => break,
+                _ = sleep(sleep_duration) => (),
+            }
+        }
+
+        Ok(())
+    }
+
+    async fn last_rewarded_end_time(&self) -> db_store::Result<DateTime<Utc>> {
+        Utc.timestamp_opt(meta::fetch(&self.pool, "last_rewarded_end_time").await?, 0)
+            .single()
+            .ok_or(db_store::Error::DecodeError)
+    }
+
+    async fn next_rewarded_end_time(&self) -> db_store::Result<DateTime<Utc>> {
+        Utc.timestamp_opt(meta::fetch(&self.pool, "next_rewarded_end_time").await?, 0)
+            .single()
+            .ok_or(db_store::Error::DecodeError)
+    }
+
+    async fn disable_complete_data_checks_until(&self) -> db_store::Result<DateTime<Utc>> {
+        Utc.timestamp_opt(
+            meta::fetch(&self.pool, "disable_complete_data_checks_until").await?,
+            0,
+        )
+        .single()
+        .ok_or(db_store::Error::DecodeError)
+    }
+
+    pub async fn is_data_current(
+        &self,
+        reward_period: &Range<DateTime<Utc>>,
+    ) -> anyhow::Result<bool> {
+        // Check if we have heartbeats and speedtests past the end of the reward period
+        if reward_period.end >= self.disable_complete_data_checks_until().await? {
+            if sqlx::query_scalar::<_, i64>(
+                "SELECT COUNT(*) FROM heartbeats WHERE latest_timestamp >= $1",
```

Comment:
are we guaranteed that when these checks are being run there are no records being written to the database for the _next_ reward period? and if we haven't exhausted the results for the reward period when the rewarder first runs and it goes to sleep we won't pick up any records for the following period during that time?